probert/multipath.py

```diff
@@ -21,41 +21,48 @@
                              'target_wwnn', 'host_wwpn', 'target_wwpn',
                              'host_adapter'))
 MMap = namedtuple("MMap", ('multipath', 'sysfs', 'paths'))
+MPATH_SHOW = {
+    'paths': MPath,
+    'maps': MMap,
+}
+
 log = logging.getLogger('probert.multipath')
 
 
-def multipath_show_paths():
-    path_format = "%d %z %m %N %n %R %r %a"
-    cmd = ['multipathd', 'show', 'paths', 'raw', 'format', path_format]
+def _extract_mpath_data(cmd, show_verb):
     try:
         result = subprocess.run(cmd, stdout=subprocess.PIPE,
                                 stderr=subprocess.DEVNULL)
     except (subprocess.CalledProcessError, FileNotFoundError):
         return {}
 
+    mptype = MPATH_SHOW[show_verb]
     data = result.stdout.decode('utf-8')
-    paths = []
+    result = []
     for line in data.splitlines():
-        paths.append(MPath(*line.split())._asdict())
+        mp_dict = None
+        try:
+            mp_dict = mptype(*line.split())._asdict()
+        except TypeError as e:
+            log.debug(
+                'Failed to parse multipath %s entry: %s: %s' % (show_verb,
+                                                                line, e))
+        if mp_dict:
+            result.append(mp_dict)
+
+    return result
```

Comment:
here we return a list.

this is still the same function, right? shouldn't it always return the same type of a thing? ie. either list or dict, but not both?